server/src/main/scala/org/http4s/server/middleware/authentication/DigestAuth.scala

```diff
@@ -62,115 +74,153 @@ object DigestAuth {
     *                       purposes anymore).
     * @param nonceBits The number of random bits a nonce should consist of.
     */
-  def apply[F[_]: Sync, A](
+  def applyF[F[_]: Concurrent: Timer, A](
       realm: String,
       store: AuthenticationStore[F, A],
       nonceCleanupInterval: Duration = 1.hour,
       nonceStaleTime: Duration = 1.hour,
       nonceBits: Int = 160,
-  ): AuthMiddleware[F, A] = {
-    val nonceKeeper =
-      new NonceKeeper(nonceStaleTime.toMillis, nonceCleanupInterval.toMillis, nonceBits)
-    challenged(challenge(realm, store, nonceKeeper))
-  }
+  ): F[AuthMiddleware[F, A]] =
+    challenge[F, A](
+      realm = realm,
+      store = store,
+      nonceCleanupInterval = nonceCleanupInterval,
+      nonceStaleTime = nonceStaleTime,
+      nonceBits = nonceBits,
+    ).map { runChallenge =>
+      challenged(runChallenge)
+    }
 
-  /** Side-effect of running the returned task: If req contains a valid
-    * AuthorizationHeader, the corresponding nonce counter (nc) is increased.
-    */
   def challenge[F[_], A](realm: String, store: AuthenticationStore[F, A], nonceKeeper: NonceKeeper)(
       implicit F: Sync[F]
+  ): Kleisli[F, Request[F], Either[Challenge, AuthedRequest[F, A]]] =
+    challengeInterop[F, A](
+      realm,
+      store,
+      F.delay(nonceKeeper.newNonce()),
+      (data, nc) => F.delay(nonceKeeper.receiveNonce(data, nc)),
+    )
+
+  /** Similar to [[apply]], but exposing the underlying [[challenge]]
+    * [[cats.data.Kleisli]] instead of an entire [[AuthMiddleware]]
+    *
+    * Side-effect of running the returned task: If req contains a valid
+    * AuthorizationHeader, the corresponding nonce counter (nc) is increased.
```

Comment:
```suggestion
    * Side-effect of the middleware: If req contains a valid
    * AuthorizationHeader, the corresponding nonce counter (nc) is increased.
```

Reply:
Critically, this _isn't_ an `AuthMiddleware`, it's just a `Kleisli` -- this is more of a proto-middleware, do you still feel strongly about this change?

Reply:
I'm sorry, you're completely right, I was mixing up my terminology. A middleware is not a `Kleisli` but a `Kleisli => Kleisli`. Ignore me :)